huaweicloud/services/rds/data_source_huaweicloud_rds_engine.go

```diff
@@ -0,0 +1,109 @@
+package rds
+
+import (
+	"context"
+	"sort"
+
+	"github.com/chnsz/golangsdk/openstack/rds/v3/instances"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/helper/hashcode"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/fmtp"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/logp"
+)
+
+var (
+	// Database type of MySQL
+	DBMySQL = "MySQL"
+	// Database type of Postgre SQL
+	DBPostgreSQL = "PostgreSQL"
+	// Database type of Microsoft SQL Server
+	DBSQLServer = "SQLServer"
+)
+
+type Versions []instances.VersionInfo
+
+func (this Versions) Len() int {
+	return len(this)
+}
+
+func (this Versions) Less(i, j int) bool {
+	return this[i].Name < this[j].Name
+}
+
+func (this Versions) Swap(i, j int) {
+	this[i], this[j] = this[j], this[i]
+}
+
+func DataSourceRdsEngineV3() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceRdsEngineV3Read,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					DBMySQL, DBPostgreSQL, DBSQLServer,
+				}, false),
+				Default: DBMySQL,
+			},
+			"versions": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceRdsEngineV3Read(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.RdsV3Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating Huaweicloud RDS v3 client: %s", err)
+	}
+
+	engineType := d.Get("type").(string)
+	engine, err := instances.ListEngine(client, engineType)
+	if err != nil {
+		return fmtp.DiagErrorf("Error getting database engine list: %s", err)
+	}
+
+	versions := engine.Versions
+	sort.Sort(Versions(versions))
```

Comment:
recommended to use `sort.Slice`, no need to define `Len`, `Less` and `Swap`
```
sort.Slice(versions, func(i, j int) bool {
	return versions[i].Name < versions[j].Name
})
```

Reply:
good recommendation.